### DISCUSSION TITLE
Psono Import doesn't import all Passwords

### DISCUSSION THREAD

**Discussion by Ryeera** [dani-garcia/vaultwarden#4366] [Q&A]
### Missing passwords from import
First of all, LOVE the project, thank you so much!
I wanted to properly evaluate if Vaultwarden would be a valid replacement to Psono for my family and me since Psono is not a good solution for us (Webauthn not working on phones to name one issue).
To properly evaluate, I wanted to import my passwords form Psono and was very happy to see that there's an option for that! However, once imported, I noticed that out of my over 400 passwords, only about 10% were imported. See my further analysis below.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.30.3
* Web-vault version: v2024.1.2
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.44.0
* Clients used: Web-Client
* Reverse proxy and version: NPM 2.10.4
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*****************",
  "domain_origin": "*****://*****************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "experimental_client_feature_flags": "fido2-vault-credentials",
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/vaultwarden.log",
  "log_level": "warn",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_identity_uri": "https://identity.bitwarden.eu",
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.eu",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "***********,**********",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "****************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**********",
  "smtp_password": "***",
  "smtp_port": 465,
  "smtp_security": "force_tls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "******************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "user_send_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": "*****",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

### Steps to reproduce
1. Export passwords from Psono in unencrypted JSON-format
2. Import passwords into Vaultwarden, tried both file and pasted json

### Expected behaviour
All passwords should be imported into their corresponding top-level folders. I would expect passwords in nested folders to be imported into the top-level folder. So a password sitting in /folder1/subfolder1/ would be imported into folder1.

### Actual behaviour
Only passwords in the root-folder or only one folder in are imported. All passwords residing in sub-folders are ignored. Additionally for passwords in folders, only the first one is put into the correct folder, the rest is imported without folder.

### Troubleshooting data
Psono with Family-passwords showing (obfuscated of course), folder is 16 passwords:
<img width="245" alt="image" src="https://github.com/dani-garcia/vaultwarden/assets/79052127/9dd3ff8c-dd01-41e0-87dd-de49945d4657">

Import Result:
<img width="346" alt="image" src="https://github.com/dani-garcia/vaultwarden/assets/79052127/4a49d490-2229-4c24-88fb-a6f5e7d30bcd">

Family-Folder in Vaultwarden:
<img width="502" alt="image" src="https://github.com/dani-garcia/vaultwarden/assets/79052127/fe118c4b-d641-4947-b4d6-b7c6887d46b0">

**Comment by BlackDex** [comment]
This is not something this project can fix. Imports are done fully by the clients. They convert the format and send that to the backend.

Best would be to check the issues at the Bitwarden repo or community forum.

**Reply by Ryeera** [reply]
I'm using the website which is provided by vaultwarden. Is that just a complete clone of the Bitwarden Web-Client?

**Reply by BlackDex** [reply]
It's a clone with a few minor patches.
The same behavior will happen in the SaaS solution

**Reply by Ryeera** [reply]
Okay I just created an account on Bitwarden just to check this and this is not true. I could import all 442 items into my bitwarden-account plus it created nested folders even! I thought these weren't possible, but they are, even in Vaultwarden.
So something is definitely going wrong here. I can't find the code for the Psono-Import in this repository, so I can't check it, but since it is working just fine in the official Web-App, I'd say there must be an error somewhere in Vaultwarden.

**Reply by Ryeera** [reply]
Just tried importing into VaultWarden using the Bitwarden Chrome Extension and that worked perfectly! I can live with this ^^

**Reply by BlackDex** [reply]
Probably you are using a newer client as extension then the web-vault currently is.
Which would indicate that a newer web-vault would solve the issue too.

**Reply by Ryeera** [reply]
Can I update the web-client myself or do I need to wait for a new release? I mean to be fair, either is fine, since I could import the passwords now and don't really need to import passwords every other day ^^"
